### PR TITLE
conformance 015

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**

### CODE REVIEW

File: src/protocol/payload/block.rs [hunk 1]
```diff
@@ -174,17 +183,15 @@ impl Codec for Header {
 
 impl Header {
     /// Calculates the double Sha256 hash for [Header]
-    pub fn double_sha256(&self) -> std::io::Result<Hash> {
+    pub fn double_sha256(&self) -> Hash {
         let mut buffer = Vec::new();
 
-        self.encode_without_tx_count(&mut buffer)?;
+        self.encode_without_tx_count(&mut buffer).unwrap();
```
Comment (niklaslong) [inline src/protocol/payload/block.rs:189]: I'm not convinced we'd want to unwrap this? In general having protocol implementation code return errors would be better, we can always unwrap at the test level.

File: src/protocol/payload/tx.rs [hunk 2]
```diff
@@ -21,17 +21,15 @@ pub enum Tx {
 }
 
 impl Tx {
-    pub fn double_sha256(&self) -> std::io::Result<Hash> {
+    pub fn double_sha256(&self) -> Hash {
         let mut buffer = Vec::new();
 
-        self.encode(&mut buffer)?;
+        self.encode(&mut buffer).unwrap();
```
Comment (niklaslong) [inline src/protocol/payload/tx.rs:27]: Same unwrap considerations as above.

File: src/tests/conformance/messages.rs [hunk 3]
```diff
@@ -378,6 +378,116 @@ async fn correctly_lists_peers() {
...
+    //  3. For each test case:
+    //      a) send GetBlocks
+    //      b) receive Inv
+    //      c) assert Inv received matches expectations
+    //
+    // The test currently fails for both Zebra and zcashd.
+    //
+    // Current behaviour:
+    //
+    //  zcashd: Fails because it ignores requests when the last hash is the last hash it knows of. We expect it to return
```
Comment (niklaslong) [inline src/tests/conformance/messages.rs:404]: Wouldn't this be expected with `start < range < stop` as opposed to `<= stop`?

File: src/tests/conformance/messages.rs [hunk 4]
```diff
@@ -378,6 +378,116 @@ async fn correctly_lists_peers() {
...
+    .log_to_stdout(true)
+    .start()
+    .await;
+
+    let blocks = Block::initial_testnet_blocks();
+
+    let mut stream = initiate_handshake(node.addr()).await.unwrap();
+    let filter = MessageFilter::with_all_auto_reply();
+
+    // Test GetBlocks(i) -> Inv(i+1..)
```
Comment (niklaslong) [inline src/tests/conformance/messages.rs:429]: This makes sense to an informed reader but it would be good to have some more context in plain english?

File: src/tests/conformance/messages.rs [hunk 5]
```diff
@@ -378,6 +378,116 @@ async fn correctly_lists_peers() {
...
+            vec![block.double_sha256()],
+            Hash::zeroed(),
+        ))
+        .write_to_stream(&mut stream)
+        .await
+        .unwrap();
+
+        match filter.read_from_stream(&mut stream).await.unwrap() {
+            Message::Inv(inv) => {
+                let inv_hashes = blocks.iter().skip(i + 1).map(|b| b.inv_hash()).collect();
```
Comment (niklaslong) [inline src/tests/conformance/messages.rs:441]: In conjunction with the above, the `i + 1` detail can easily be overlooked.

File: src/tests/conformance/messages.rs [hunk 6]
```diff
@@ -378,6 +378,116 @@ async fn correctly_lists_peers() {
...
+        Message::Inv(inv) => {
+            let expected = Inv::new(vec![blocks[1].inv_hash()]);
+            assert_eq!(inv, expected);
+        }
+        message => panic!("Expected Inv, but got {:?}", message),
+    }
+
+    // Test that we get corrected if we are "off chain".
+    // We expect that unknown hashes get ignored, until it finds a known hash; it then returns
+    // all known children of that block.
```
Comment (niklaslong) [inline src/tests/conformance/messages.rs:467]: Isn't it also supposed to send a `NotFound` in conjunction with the `Inv`?

**Review by niklaslong — CHANGES_REQUESTED** [review]
Great to have this implemented 🎉 
I left a few comments, most notably about limiting the use of `unwrap` in setup and protocol implementation code.

File: src/tests/conformance/messages.rs (hunk 3 above)
Comment (Mirko-von-Leipzig) [reply src/tests/conformance/messages.rs:404]: I would still expect a reply; even if its just the empty set `Inv({})`. 

From the spec:
"If a block hash is found in the node's main chain, the list of its children is returned back via the inv message" - in this case the list is simply empty.

File: src/tests/conformance/messages.rs (hunk 4 above)
Comment (Mirko-von-Leipzig) [reply src/tests/conformance/messages.rs:429]: Good point

File: src/protocol/payload/block.rs (hunk 1 above)
Comment (Mirko-von-Leipzig) [reply src/protocol/payload/block.rs]: Reverted
